File: rnamake/unittests/motif_state_ensemble_tree_unittest.py
import unittest
import os
import random
import rnamake.resource_manager as rm
import rnamake.motif_tree as motif_tree

from rnamake import sqlite_library
from rnamake import motif_graph, motif_topology, motif_state_ensemble_tree

class MotifStateEnsembleTreeUnittest(unittest.TestCase):

    def test_creation(self):
        mset = motif_state_ensemble_tree.MotifStateEnsembleTree()

    def test_add(self):
        mset = motif_state_ensemble_tree.MotifStateEnsembleTree()
        mse = rm.manager.get_motif_state_ensemble(name="GG_LL_CC_RR")
        mset.add_ensemble(mse)
        mset.add_ensemble(mse)
        if len(mset) != 2:
            self.fail("did not properly add ensembles")

    def test_to_mst(self):
        mset = motif_state_ensemble_tree.MotifStateEnsembleTree()
        mse = rm.manager.get_motif_state_ensemble(name="GG_LL_CC_RR")
        mset.add_ensemble(mse)
        mset.add_ensemble(mse)
        mset.add_ensemble(mse)
        mset.add_ensemble(mse)

        mst = mset.to_mst()
        #print mst.get_node(0).data.name()
        #mst.write_pdbs()

    def _test_enumerator(self):
        lib = sqlite_library.MotifStateEnsembleSqliteLibrary("all_bp_steps")
        mtst = motif_state_ensemble_tree.MotifStateEnsembleTree()
        mtst.add_ensemble(lib.get(name="GG_LL_CC_RR"))
        mtst.add_ensemble(lib.get(name="GG_LL_CC_RR"))
        enumerator = motif_state_ensemble_tree.MotifStateEnsembleTreeEnumerator(mtst)
        enumerator.record()

    def test_setup_from_mt_4(self):
        mg = motif_graph.MotifGraph()
        mg.add_motif(m_name="HELIX.IDEAL.20")
        mg.replace_ideal_helices()
        mt = motif_topology.graph_to_tree(mg)
        mset = motif_state_ensemble_tree.MotifStateEnsembleTree(mt)

        mst = mset.to_mst()
        #for r in mt.residues():
        #    print mst.get_residue(r.uuid)

    def test_motif_sub(self):
        mlib = sqlite_library.MotifSqliteLibrary("twoway")
        m1 = mlib.get_random()
        m2 = mlib.get_random()
        while len(m1.sequence()) == len(m2.sequence()):
            m2 = mlib.get_random()

        motif_state_ensemble_tree.build_motif_sub_for_motif_state_ensemble(m1, [m2], mlib)
        rm.manager.register_extra_motif_ensembles("test.dat")

        r = rm.manager.has_supplied_motif_ensemble(m1.name, m1.ends[0].name())
        self.failUnless(r == 1, "did not find sub")

        r = rm.manager.has_supplied_motif_ensemble(m1.name, m1.ends[1].name())
        self.failUnless(r == 1, "did not find sub")

        mt = motif_tree.MotifTree()
        mt.add_motif(m_name="HELIX.IDEAL")
        mt.add_motif(m1)
        mt.add_motif(m_name="HELIX.IDEAL")

        mset = motif_state_ensemble_tree.MotifStateEnsembleTree(mt=mt)
        mst = mset.to_mst()

        self.failUnless(mst.get_node(1).data.cur_state.name == m2.name)

        os.remove("test.dat")


def main():
    unittest.main()

if __name__ == '__main__':
    main()

File: rnamake/lib/RNAMake/unittests/motif_data_structures_unittests/motif_state_ensemble_tree_unittest.cpp
//
//  motif_state_ensemble_tree_unittest.cpp
//  RNAMake
//
//  Created by Joseph Yesselman on 9/3/15.
//  Copyright (c) 2015 Joseph Yesselman. All rights reserved.
//

#include "build_motif_tree.h"
#include "motif_state_ensemble_tree_unittest.h"
#include "motif_data_structures/motif_state_ensemble_tree.h"
#include "resources/resource_manager.h"
#include "resources/motif_state_ensemble_sqlite_library.h"


int
MotifStateEnsembleTreeUnittest::test_creation() {
    MotifStateEnsembleTree mse_tree;
    return 1;
}

int
MotifStateEnsembleTreeUnittest::test_add_ensemble() {
    MotifStateEnsembleTree mse_tree;
    auto mse = ResourceManager::getInstance().get_motif_state_ensemble("GG_LL_CC_RR");
    mse_tree.add_ensemble(mse);
    mse_tree.add_ensemble(mse);
    if(mse_tree.size() != 2) { return 0; }
    return 1;
}

int
MotifStateEnsembleTreeUnittest::test_to_mst() {
    MotifStateEnsembleTree mset;
    auto mse = ResourceManager::getInstance().get_motif_state_ensemble("GG_LL_CC_RR");
    for(int i = 0; i < 10; i++) { mset.add_ensemble(mse); }
    
    auto mst = mset.to_mst();
    mst->to_motif_tree()->write_pdbs();
    return 1;
    
}

int
MotifStateEnsembleTreeUnittest::test_from_mt() {
    BuildMotifTree builder(Strings{"bp_steps", "twoway"});
    auto mt = builder.build(10);
    MotifStateEnsembleTree mset;
    mset.setup_from_mt(mt);
    auto mst = mset.to_mst();
    
    return 1;
}

int
MotifStateEnsembleTreeUnittest::test_enumerator() {
    MotifStateEnsembleSqliteLibrary lib("bp_steps");
    lib.load_all();
    for(auto const & mes : lib) {
    
        auto mtst = std::make_shared<MotifStateEnsembleTree>();
        mtst->add_ensemble(mes);
        MotifStateEnsembleTreeEnumerator enumerator(mtst);
        enumerator.record(mes->id());
    }
    
    return 1;
}


int
MotifStateEnsembleTreeUnittest::run() {
    //if (test_creation() == 0)      { std::cout << "test_creation failed" << std::endl;  }
    //if (test_add_ensemble() == 0)  { std::cout << "test_add_ensemble failed" << std::endl;  }
    //if (test_to_mst() == 0)        { std::cout << "test_to_mst failed" << std::endl;  }
    //if (test_from_mt() == 0)       { std::cout << "test_from_mt failed" << std::endl;  }
    if (test_enumerator() == 0)      { std::cout << "test_enumerator failed" << std::endl;  }

    return 1;
}